packages/api/src/folder/service.ts

```diff
@@ -0,0 +1,106 @@
+import createError from "http-errors";
+import { logger } from "@stela/logger";
+import { db } from "../database";
+import type {
+  Folder,
+  FolderColumnsForUpdate,
+  PatchFolderRequest,
+} from "./models";
+import { requestFieldsToDatabaseFields } from "./helper";
+import { permission } from "../access/permission";
+import { folderAccess } from "./access";
+import { accountService } from "../account/service";
+
+export const getFolderById = async (requestQuery: {
+  folderId: string;
+}): Promise<Folder> => {
+  const folder = await db
+    .sql<Folder>("folder.queries.get_folder_by_id", {
+      folderId: requestQuery.folderId,
+    })
+    .catch((err) => {
+      logger.error(err);
+      throw new createError.InternalServerError("Failed to retrieve folder");
+    });
+
+  if (folder.rows[0] === undefined) {
+    throw new createError.NotFound("Folder not found.");
+  }
+
+  return folder.rows[0];
+};
+
+function buildPatchQuery(patchFolderRequest: FolderColumnsForUpdate): string {
+  const updates = Object.entries(patchFolderRequest)
+    .filter(([key, value]) => value !== undefined && key !== "folderId")
+    .map(([key, _]) => `${key} = :${key}`);
+
+  if (updates.length === 0) {
+    throw new createError.BadRequest("Request cannot be empty");
+  }
+
+  const query = `
+    UPDATE folder
+    SET ${updates.join(", ")}
+    WHERE folderid = :folderId
+    RETURNING folder.folderid AS "folderId"
+  `;
+
+  return query.trim();
+}
+
+const validateCanPatchFolder = async (
+  folderId: string,
+  emailFromAuthToken: string
+): Promise<void> => {
+  const folder = await getFolderById({ folderId });
+  let canEdit = false;
+  const accountArchive = await accountService.getAccountArchive(
+    folder.archiveId,
+    emailFromAuthToken
+  );
+  if (
+    accountArchive &&
+    (await permission.checkCanEditAccountArchive(accountArchive))
+  ) {
+    canEdit = true;
+  }
+  if (!canEdit) {
+    const currentAccountArchive = await accountService.getCurrentAccountArchive(
```

Comment:
Remember that we also need to check the access role on the `account_archive` for the archive that received a share; if the user only has viewer permissions on the archive they shouldn't be able to edit a share with that archive regardless of the permissions of the share itself.